File: backend/add_npcl_data.js
const fs = require('fs');
const path = require('path');

const chargesFile = '/Users/yashgupta/IEX-Dashboard/backend_tables_updated - state_charges.csv';
const tariffFile = '/Users/yashgupta/IEX-Dashboard/backend_tables_updated - state-tarriff.csv';

// 1. Update State Charges
let chargesCsv = fs.readFileSync(chargesFile, 'utf8').trim().split('\n');
let chargesHeader = chargesCsv[0].split(',');
let hasDiscom = chargesHeader[1] === 'discom';

if (!hasDiscom) {
    chargesHeader.splice(1, 0, 'discom');
    chargesCsv[0] = chargesHeader.join(',');
    for (let i = 1; i < chargesCsv.length; i++) {
        if (chargesCsv[i].trim()) {
            let row = chargesCsv[i].split(',');
            row.splice(1, 0, '');
            chargesCsv[i] = row.join(',');
        }
    }
} else {
    // Remove existing NPCL rows
    chargesCsv = chargesCsv.filter((line, i) => i === 0 || (line.split(',')[1] !== 'NPCL'));
}

const addCharge = (cat, sub_cat, supp_cat, volt, from, to, fixed, cross, wheel, stu, stuloss, wheelloss) => {
    // state,discom,category,sub_category,supply_voltage_category,voltage_level,from_date,to_date,demand_fixed_charge_kva_per_month_Rs,cross_subsidy,distribution_wheeling_charges,stu_charges,stu_loss_percent,wheeling_loss_percent,additional_charge
    chargesCsv.push(`UTTAR_PRADESH,NPCL,${cat},${sub_cat},${supp_cat},${volt},${from},${to},${fixed},${cross},${wheel},${stu},${stuloss},${wheelloss},0`);
}

addCharge('HV-1', '', 'High Tension (HT)', '11', '01-04-2024', '31-03-2025', 430, 1.33, 1.012, 0.2326, 3.18, 5.66);
addCharge('HV-1', '', 'High Tension (HT)', '33', '01-04-2024', '31-03-2025', 400, 1.40, 1.012, 0.2326, 3.18, 2.15);
addCharge('HV-2', '', 'High Tension (HT)', '11', '01-04-2024', '31-03-2025', 300, 0.57, 1.012, 0.2326, 3.18, 5.66);
addCharge('HV-2', '', 'High Tension (HT)', '33', '01-04-2024', '31-03-2025', 290, 0.59, 1.012, 0.2326, 3.18, 2.15);

addCharge('HV-1', '', 'High Tension (HT)', '11', '01-04-2025', '31-03-2026', 430, 1.33, 0.9985, 0.3075, 3.18, 5.47);
addCharge('HV-1', '', 'High Tension (HT)', '33', '01-04-2025', '31-03-2026', 400, 1.40, 0.9985, 0.3075, 3.18, 0.88);
addCharge('HV-2', '', 'High Tension (HT)', '11', '01-04-2025', '31-03-2026', 300, 0.57, 0.9985, 0.3075, 3.18, 5.47);
addCharge('HV-2', '', 'High Tension (HT)', '33', '01-04-2025', '31-03-2026', 290, 0.59, 0.9985, 0.3075, 3.18, 0.88);

addCharge('HV-1', '', 'High Tension (HT)', '11', '01-04-2026', '31-03-2027', 430, 1.33, 1.03, 0.3075, 3.18, 2.58);
addCharge('HV-1', '', 'High Tension (HT)', '33', '01-04-2026', '31-03-2027', 400, 1.40, 1.03, 0.3075, 3.18, 0.79);
addCharge('HV-2', '', 'High Tension (HT)', '11', '01-04-2026', '31-03-2027', 300, 0.57, 1.03, 0.3075, 3.18, 2.58);
addCharge('HV-2', '', 'High Tension (HT)', '33', '01-04-2026', '31-03-2027', 290, 0.59, 1.03, 0.3075, 3.18, 0.79);

fs.writeFileSync(chargesFile, chargesCsv.join('\n'));

// 2. Update State Tariff
let tariffCsv = fs.readFileSync(tariffFile, 'utf8').trim().split('\n');
let tariffHeader = tariffCsv[0].split(',');
let hasTariffDiscom = tariffHeader[1] === 'discom';

if (!hasTariffDiscom) {
    tariffHeader.splice(1, 0, 'discom');
    tariffCsv[0] = tariffHeader.join(',');
    for (let i = 1; i < tariffCsv.length; i++) {
        if (tariffCsv[i].trim()) {
            let row = tariffCsv[i].split(',');
            row.splice(1, 0, '');
            tariffCsv[i] = row.join(',');
        }
    }
} else {
    tariffCsv = tariffCsv.filter((line, i) => i === 0 || (line.split(',')[1] !== 'NPCL'));
}

const addTariff = (cat, sub_cat, supp_cat, volt, month, start, end, base, unit, percent, energy) => {
    // state,discom,consumer_category,sub_category,supply_voltage_category,supply_voltage,month,tod_start_time,tod_end_time,base_energy_rate,base_energy_unit,tod_charge_percent,energy_rate
    tariffCsv.push(`UTTAR_PRADESH,NPCL,${cat},${sub_cat},${supp_cat},${volt},${month},${start},${end},${base},${unit},${percent},${energy}`);
}

for (let month = 1; month <= 12; month++) {
    let is_summer = month >= 4 && month <= 9;
    
    // HV-1 11kV
    addTariff('HV-1', '', 'High Tension (HT)', '11', month, '07:00', '16:00', 8.32, 'kWh', 0, 8.32);
    addTariff('HV-1', '', 'High Tension (HT)', '11', month, '16:00', '19:00', 8.32, 'kWh', 0, 8.32);
    addTariff('HV-1', '', 'High Tension (HT)', '11', month, '19:00', '02:00', 8.32, 'kWh', 0, 8.32);
    addTariff('HV-1', '', 'High Tension (HT)', '11', month, '02:00', '07:00', 8.32, 'kWh', 0, 8.32);

    // HV-1 33kV
    addTariff('HV-1', '', 'High Tension (HT)', '33', month, '07:00', '16:00', 8.12, 'kWh', 0, 8.12);
    addTariff('HV-1', '', 'High Tension (HT)', '33', month, '16:00', '19:00', 8.12, 'kWh', 0, 8.12);
    addTariff('HV-1', '', 'High Tension (HT)', '33', month, '19:00', '02:00', 8.12, 'kWh', 0, 8.12);
    addTariff('HV-1', '', 'High Tension (HT)', '33', month, '02:00', '07:00', 8.12, 'kWh', 0, 8.12);

    // HV-2 11kV
    if (is_summer) {
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '07:00', '16:00', 6.03, 'kWh', 0, 6.03);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '16:00', '19:00', 7.10, 'kWh', 0, 7.10);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '19:00', '02:00', 8.17, 'kWh', 0, 8.17);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '02:00', '07:00', 8.17, 'kWh', 0, 8.17);
    } else {
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '10:00', '17:00', 7.10, 'kWh', 0, 7.10);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '17:00', '19:00', 8.165, 'kWh', 0, 8.165);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '19:00', '22:00', 7.10, 'kWh', 0, 7.10);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '22:00', '04:00', 6.035, 'kWh', 0, 6.035);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '04:00', '06:00', 7.10, 'kWh', 0, 7.10);
        addTariff('HV-2', '', 'High Tension (HT)', '11', month, '06:00', '10:00', 8.165, 'kWh', 0, 8.165);
    }

    // HV-2 33kV
    if (is_summer) {
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '07:00', '16:00', 5.78, 'kWh', 0, 5.78);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '16:00', '19:00', 6.80, 'kWh', 0, 6.80);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '19:00', '02:00', 7.82, 'kWh', 0, 7.82);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '02:00', '07:00', 7.82, 'kWh', 0, 7.82);
    } else {
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '10:00', '17:00', 6.80, 'kWh', 0, 6.80);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '17:00', '19:00', 7.82, 'kWh', 0, 7.82);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '19:00', '22:00', 6.80, 'kWh', 0, 6.80);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '22:00', '04:00', 5.78, 'kWh', 0, 5.78);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '04:00', '06:00', 6.80, 'kWh', 0, 6.80);
        addTariff('HV-2', '', 'High Tension (HT)', '33', month, '06:00', '10:00', 7.82, 'kWh', 0, 7.82);
    }
}

fs.writeFileSync(tariffFile, tariffCsv.join('\n'));

console.log("CSVs updated successfully with JS!");

File: backend/add_npcl_data.py
import pandas as pd
import numpy as np

# 1. Update State Charges
df_charges = pd.read_csv('/Users/yashgupta/IEX-Dashboard/backend_tables_updated - state_charges.csv')

# Insert discom column after state if it doesn't exist
if 'discom' not in df_charges.columns:
    df_charges.insert(1, 'discom', '')

# Remove existing NPCL data if any (just in case)
df_charges = df_charges[df_charges['discom'] != 'NPCL']

new_charges = []

def add_charge(category, supply_voltage_category, voltage_level, from_date, to_date, fixed, cross, wheeling, stu, stu_loss, wheeling_loss):
    new_charges.append({
        'state': 'UTTAR_PRADESH',
        'discom': 'NPCL',
        'category': category,
        'sub_category': '',
        'supply_voltage_category': supply_voltage_category,
        'voltage_level': voltage_level,
        'from_date': from_date,
        'to_date': to_date,
        'demand_fixed_charge_kva_per_month_Rs': fixed,
        'cross_subsidy': cross,
        'distribution_wheeling_charges': wheeling,
        'stu_charges': stu,
        'stu_loss_percent': stu_loss,
        'wheeling_loss_percent': wheeling_loss,
        'additional_charge': 0
    })

# Add for FY 24-25 (01-04-2024 to 31-03-2025)
add_charge('HV-1', 'High Tension (HT)', '11', '01-04-2024', '31-03-2025', 430, 1.33, 1.012, 0.2326, 3.18, 5.66)
add_charge('HV-1', 'High Tension (HT)', '33', '01-04-2024', '31-03-2025', 400, 1.40, 1.012, 0.2326, 3.18, 2.15)
add_charge('HV-2', 'High Tension (HT)', '11', '01-04-2024', '31-03-2025', 300, 0.57, 1.012, 0.2326, 3.18, 5.66)
add_charge('HV-2', 'High Tension (HT)', '33', '01-04-2024', '31-03-2025', 290, 0.59, 1.012, 0.2326, 3.18, 2.15)

# Add for FY 25-26 (01-04-2025 to 31-03-2026)
add_charge('HV-1', 'High Tension (HT)', '11', '01-04-2025', '31-03-2026', 430, 1.33, 0.9985, 0.3075, 3.18, 5.47)
add_charge('HV-1', 'High Tension (HT)', '33', '01-04-2025', '31-03-2026', 400, 1.40, 0.9985, 0.3075, 3.18, 0.88)
add_charge('HV-2', 'High Tension (HT)', '11', '01-04-2025', '31-03-2026', 300, 0.57, 0.9985, 0.3075, 3.18, 5.47)
add_charge('HV-2', 'High Tension (HT)', '33', '01-04-2025', '31-03-2026', 290, 0.59, 0.9985, 0.3075, 3.18, 0.88)

# Add for FY 26-27 (01-04-2026 to 31-03-2027)
add_charge('HV-1', 'High Tension (HT)', '11', '01-04-2026', '31-03-2027', 430, 1.33, 1.03, 0.3075, 3.18, 2.58)
add_charge('HV-1', 'High Tension (HT)', '33', '01-04-2026', '31-03-2027', 400, 1.40, 1.03, 0.3075, 3.18, 0.79)
add_charge('HV-2', 'High Tension (HT)', '11', '01-04-2026', '31-03-2027', 300, 0.57, 1.03, 0.3075, 3.18, 2.58)
add_charge('HV-2', 'High Tension (HT)', '33', '01-04-2026', '31-03-2027', 290, 0.59, 1.03, 0.3075, 3.18, 0.79)

df_charges = pd.concat([df_charges, pd.DataFrame(new_charges)], ignore_index=True)
df_charges.to_csv('/Users/yashgupta/IEX-Dashboard/backend_tables_updated - state_charges.csv', index=False)


# 2. Update State Tariff
df_tariff = pd.read_csv('/Users/yashgupta/IEX-Dashboard/backend_tables_updated - state-tarriff.csv')

if 'discom' not in df_tariff.columns:
    df_tariff.insert(1, 'discom', '')

df_tariff = df_tariff[df_tariff['discom'] != 'NPCL']

new_tariffs = []
def add_tariff(category, voltage, month, start, end, rate):
    new_tariffs.append({
        'state': 'UTTAR_PRADESH',
        'discom': 'NPCL',
        'consumer_category': category,
        'sub_category': '',
        'supply_voltage_category': 'High Tension (HT)',
        'supply_voltage': voltage,
        'month': month,
        'tod_start_time': start,
        'tod_end_time': end,
        'base_energy_rate': rate,
        'base_energy_unit': 'kWh',
        'tod_charge_percent': 0,
        'energy_rate': rate
    })

# The energy charges for NPCL are the same for all years (FY24-25, 25-26, 26-27 in the screenshot)
for month in range(1, 13):
    is_summer = month >= 4 and month <= 9
    
    # HV-1 (11 kV)
    for tod, rate in [('07:00', '16:00', 8.32), ('16:00', '19:00', 8.32), ('19:00', '02:00', 8.32), ('02:00', '07:00', 8.32)]:
        add_tariff('HV-1', '11', month, tod[0], tod[1], rate)
    
    # HV-1 (33 kV)
    for tod, rate in [('07:00', '16:00', 8.12), ('16:00', '19:00', 8.12), ('19:00', '02:00', 8.12), ('02:00', '07:00', 8.12)]:
        add_tariff('HV-1', '33', month, tod[0], tod[1], rate)

    # HV-2 (11 kV)
    if is_summer:
        tods = [('07:00', '16:00', 6.03), ('16:00', '19:00', 7.10), ('19:00', '02:00', 8.17), ('02:00', '07:00', 8.17)]
    else:
        tods = [('10:00', '17:00', 7.1), ('17:00', '19:00', 8.165), ('19:00', '22:00', 7.1), ('22:00', '04:00', 6.035), ('04:00', '06:00', 7.1), ('06:00', '10:00', 8.165)]
    for tod in tods:
        add_tariff('HV-2', '11', month, tod[0], tod[1], tod[2])
        
    # HV-2 (33 kV)
    if is_summer:
        tods = [('07:00', '16:00', 5.78), ('16:00', '19:00', 6.80), ('19:00', '02:00', 7.82), ('02:00', '07:00', 7.82)]
    else:
        tods = [('10:00', '17:00', 6.8), ('17:00', '19:00', 7.82), ('19:00', '22:00', 6.8), ('22:00', '04:00', 5.78), ('04:00', '06:00', 6.8), ('06:00', '10:00', 7.82)]
    for tod in tods:
        add_tariff('HV-2', '33', month, tod[0], tod[1], tod[2])

df_tariff = pd.concat([df_tariff, pd.DataFrame(new_tariffs)], ignore_index=True)
df_tariff.to_csv('/Users/yashgupta/IEX-Dashboard/backend_tables_updated - state-tarriff.csv', index=False)

print("CSVs updated successfully!")
